man/anoubis_sig.3
.\"	$OpenBSD: mdoc.template,v 1.9 2004/07/02 10:36:57 jmc Exp $
.\"
.\" Copyright (c) 2007 GeNUA mbH <info@genua.de>
.\"
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
.\" "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
.\" LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
.\" A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
.\" OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
.\" SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
.\" TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
.\" PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
.\" LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
.\" NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
.\" SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd December 17, 2008
.Dt ANOUBIS_SIG 3
.Os Anoubis
.Sh NAME
.Nm struct anoubis_sig
.Nm anoubis_sig_create
.Nm anoubis_sig_free
.Nm anoubis_sig_key2char
.Nm anoubis_sig_sign_csum
.Nm anoubisd_sig_verify_csum
.Nm anoubis_sig_sign_policy
.Nm anoubis_sig_verify_policy_file
.Nm anoubis_sig_verify_policy
.Nm anoubis_sig_cert_name
.Nd Signature handling of Anoubis
.Sh SYNOPSIS
.In anoubis_sig.h
.Ft int
.Fn anoubis_sig_create "struct anoubis_sig **asp, const char *keyfile, const char *certfile, void *passcb"
.Ft void
.Fn anoubis_sig_free "struct anoubis_sig *as"
.Ft char *
.Fn anoubis_sig_key2char "int len, unsigned char *key"
.Ft unsigned char *
.Fn anoubis_sign_csum "struct anoubis_sig *as, unsigned char csum[ANOUBIS_SIG_HASH_SHA256_LEN], unsigned int *len"
.Ft int
.Fn anoubis_csum_verify "struct anoubis_sig *as, unsigned char csum[ANOUBIS_SIG_HASH_SHA256_LEN], unsigned char *sign, int siglen"
.Ft int
.Fn anoubisd_csum_verify "EVP_PKEY *pkey, unsigned char csum[ANOUBIS_SIG_HASH_SHA256_LEN], unsigned char *sign, int siglen"
.Ft unsigned char *
.Fn anoubis_sign_policy "struct anoubis_sig *as, const char *file, unsigned int *len"
.Ft int
.Fn anoubis_verify_policy "const char *filename, unsigned char *sigbuf, int siglen, EVP_PKEY *sigkey"
.Ft int
.Fn anoubis_verify_policy_file "const char *filename, EVP_PKEY *sigkey"
.Ft char *
.Fn anoubis_sig_cert_name "X509 *cert"

.Sh DESCRIPTION
This library is used to sign and verify checksums or policies.
.Pp
The function
.Nm anoubis_sig_create
is used to load a private and the certificate.
To free an
.Nm anoubis_sig
structure create by
.Nm anoubis_sig_create
use the function
.Nm anoubis_sig_free.
.Pp
To create a signature of a calculated checksum, use
.Nm anoubis_sign_csum.
To verify the result use
.Nm anoubisd_verify_csum.
.Pp
The function
.Nm anoubis_sign_policy
provides to create a signature for a policy file. To verify this signature for
this policy
.Nm anoubis_verify_policy_file
and
.Nm anoubis_verify_policy
are used. Where
.Nm anoubis_verify_policy_file
is using the filename of the policy and
.Nm anoubis_verify_policy
a buffer which contains the signature of the policy.
.Nm anoubis_cert_name
returns the subject name of a certificate.
.Sh RETURN VALUES
.Nm anoubis_sig_create
returns zero in case of success and a negative error code in case of an error.
.Pp
.Nm anoubis_sig_key2char
returns the char representation of the key or NULL on error
.Pp
.Nm anoubis_sign_csum
returns a unsigned char buffer including the checksum and the signature of the
checksum or NULL on error.
.Pp
.Nm anoubisd_verify_csum
returns 1 if the checksum is verified or 0 on error or not matching signature.
.Pp
.Nm anoubis_sign_policy
return the signature of the policy or NULL on error.
.Pp
.Nm anoubis_verify_policy
return 1 on successful verification or 0 on error or unsuccessful verification.
.Pp
.Nm anoubis_verify_policy_file
return 1 on successful verification or 0 on error or unsuccessful verification.
.Pp
.Nm anoubis_sig_cert_name
return subject name or NULL if an error occurs.
.Sh SEE ALSO
.Xr anoubis_client 3 ,
.Xr anoubis_server 3 ,
.Sh AUTHORS
Konrad Merz <konrad.merz@genua.de>


ChangeLog
2009-11-03 - Changes between 0.9.1 and 0.9.2

New features:

  o Initial support for dazukoFS: Make sure anoubisd is unaffected by dazukoFS
    and include dazukoFS in the kernel packages.
  o Handle OS upgrades correctly even if anoubisd is restarted in the process.
    This is done by delaying the restart until after the upgrade finished.
  o Anoubisctl/sfssig/xanoubis will now also log errors to the syslog that may
    be interesting for the administrator.

Bug fixes:

  o OpenBSD kernel: Fix an OpenBSD bug that can cause crashes in nfs_reply.
  o xanoubis: Fix directory path interfering with filter/invert options.
    BUG #1168
  o anoubisd/anoubisctl: Do not leak pid filehandle to childs.
  o xanoubis: Various fixes for operating at low screen resolutions. BUG #893
  o sfssig: Fix exporting of signatures.
  o anoubisd: Fix /etc/services not being available to anoubisd childs.
  o xanoubis: Fix 'Unexpected return-code: 1' error message. BUG #1373
  o xanoubis: Fix the Status field in the rule editor being updated too
    late. BUG #1369
  o xanoubis: Display correct state in Upgrade view.
  o xanoubis: Don't try to create a rule for an unknown binary after an
    ask event.
  o xanoubis: Make large SFS error messages scrollable. BUG #1352
  o Linux kernel: Fix EBADF when writing to a raw socket. This sometimes
    prevented dhclient from working.
  o sfssig: Fix 'add' failing if file is open for writing. BUG #1342
  o Linux kernel: Make sure a new file is not created if write access was
    denied.

Other changes:

  o Send upgrade-notify events to all attached UIs. Show the upgrade dialogue
    in xanoubis when such an event is received. Also use this to get rid of
    INotify.
  o xanoubis: Give a general overview message for SFS errors.
  o xanoubis: Save table column widths in xanoubis.conf.
  o xanoubis: Adjust column witdths to window/table size. BUG #1321
  o xanoubis: Create Anoubis specific MessageBox class.
  o xanoubis: Various text improvements.
  o anoubisd: Return EXDEV instead of EPERM when rename restrictions apply.
    This makes programs like mv and patch copy the file instead. BUG #1353
  o Handle mknod calls in kernel sfs and generate SFS-write events.


2009-10-12 - Changes between 0.1.10.D023.970.D170 and 0.9.1

New features:

  o Introduce version/compatibility checks between various Anoubis
    components. Introduce version numbers for various on disk file formats.
  o Add support for updating checksums / signatures after an operating
    system upgrade. For unsigned checksums, this can be done automatically.
    For signed checksums, xanoubis displays a list with the changed files.

Bug fixes:

  o Add a missing destructor for sfssubject
  o Add missing event_add in acc_flush of libanoubischat
  o Add upgrade lists to ComTask
  o Deactivate wizard if disconnected
  o Fix flex warnings caused by ununsed symbol in code generated by flex
  o Fix list perspective in notification tab
  o Fix memory leaks in apn parser
  o Fix notification bubbles for notify-osd
  o Fix segfault in checksum comparison for new context
  o Fixed creation of user application policy
  o Length verification for checksum operations
  o Prevent looping in module SFS when in recursive mode
  o Support display of ruleset that contain borrow
  o Truncated length of error message from SFS to prevent KDE Window Manager
    crash
  o Various I18N fixes

Other changes:

  o Get rid of apn_hash_type
  o Get rid of protocol family in filtspec rules
  o Get rid of state timeout in APN
  o Implement new sfs tree format
  o Add Upgrade marker to signature files on upgrade
  o Module SFS: New implementation of list operations
  o Move Checksums into a Subject in APN
  o Performance improvement of calls to SelectedItemCount
  o Performance improvement of hasSelection in AnList
  o Provide direct access to the SFS-Browser tab
  o Remove explicit checksums in policies
  o Unify the handling of configuration options by new AnConfig-class
  o Update the kernel configs with default mmap_min_addr
  o Use TimerEvent to check for Changes in menu.lst
  o anoubisctl and xanoubis now report PACKAGE_BUILD version


src/libanoubis/anoubissig/anoubis_sig.c
/*
 * Copyright (c) 2008 GeNUA mbH <info@genua.de>
 *
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
 * TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
 * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
 * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
 * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include "anoubis_sig.h"

static int cert_keyid(unsigned char **keyid, X509 *cert);

/* This function returns a memory area filled with the checksum (csum)
 * and the signature of the checksum (sigbuf)
 *	-------------------------------------------------
 *	|	csum	|	sigbuf			|
 *	-------------------------------------------------
 * The returned len is the complete len of the memory area
 * since the len of checksum is known.
 */

unsigned char *
anoubis_sign_csum(struct anoubis_sig *as,
    unsigned char csum[ANOUBIS_SIG_HASH_SHA256_LEN], unsigned int *len)
{
	unsigned char *sigbuf = NULL;
	unsigned char *result = NULL;
	unsigned int siglen = 0;
	EVP_MD_CTX ctx;
	EVP_PKEY *pkey;

	if (as == NULL || len == NULL || csum == NULL)
		return NULL;

	pkey = as->pkey;
	siglen = EVP_PKEY_size(pkey);

	if ((sigbuf = calloc(siglen, sizeof(unsigned char))) == NULL)
		return NULL;
	EVP_MD_CTX_init(&ctx);
	if (EVP_SignInit(&ctx, as->type) == 0) {
		EVP_MD_CTX_cleanup(&ctx);
		free(sigbuf);
		return NULL;
	}
	EVP_SignUpdate(&ctx, csum, ANOUBIS_SIG_HASH_SHA256_LEN);
	if (EVP_SignFinal(&ctx, sigbuf, &siglen, pkey) == 0) {
		free(sigbuf);
		EVP_MD_CTX_cleanup(&ctx);
		return NULL;
	}
	EVP_MD_CTX_cleanup(&ctx);
	result = calloc((ANOUBIS_SIG_HASH_SHA256_LEN + siglen),
	    sizeof(unsigned char));
	if (!result) {
		*len = 0;
		return NULL;
	}

	memcpy(result, csum, ANOUBIS_SIG_HASH_SHA256_LEN);
	memcpy(result + ANOUBIS_SIG_HASH_SHA256_LEN, sigbuf, siglen);
	free(sigbuf);
	*len = siglen + ANOUBIS_SIG_HASH_SHA256_LEN;
	return result;
}

int
anoubis_sig_verify_policy_file(const char *filename, EVP_PKEY *sigkey)
{
	char		*sigfile = NULL;
	unsigned char	*sigbuf = NULL;
	int		 siglen, fd, ret;

	if (!filename || !sigkey)
		return (-1);

	if (asprintf(&sigfile, "%s.sig", filename) == -1)
		return (-1);

	if ((fd = open(sigfile, O_RDONLY)) == -1) {
		free(sigfile);
		return (-2);
	}
	free(sigfile);

	/* Read in signature */
	siglen = EVP_PKEY_size(sigkey);
	if ((sigbuf = calloc(siglen, sizeof(unsigned char))) == NULL) {
		close(fd);
		return (-3);
	}
	while ((ret = read(fd, sigbuf, siglen)) > 0) {
		if (ret == -1) {
			close(fd);
			return (-4);
		}
	}
	close(fd);

	ret = anoubis_sig_verify_policy(filename, sigbuf, siglen, sigkey);

	free(sigbuf);

	return ret;
}

int
anoubis_sig_verify_buffer(const void *buf, int buflen, const void *sig,
    int siglen, EVP_PKEY *key)
{
	EVP_MD_CTX	ctx;
	int		result;

	if (!key || !buf || !sig)
		return -EINVAL;
	if (siglen != EVP_PKEY_size(key))
		return -ERANGE;
	EVP_MD_CTX_init(&ctx);
	if (EVP_VerifyInit(&ctx, ANOUBIS_SIG_HASH_DEFAULT) == 0) {
		EVP_MD_CTX_cleanup(&ctx);
		return -ENOMEM;
	}
	EVP_VerifyUpdate(&ctx, buf, buflen);
	result = EVP_VerifyFinal(&ctx, sig, siglen, key);
	EVP_MD_CTX_cleanup(&ctx);
	if (result <= 0)
		return -EPERM;
	return 0;
}

int
anoubis_sig_sign_buffer(const void *buf, int buflen, void **sigp, int *siglenp,
    struct anoubis_sig *as)
{
	unsigned int	 len;
	int		 result;
	void		*sig;
	EVP_MD_CTX	 ctx;

	if (!as || !as->pkey || !buf || !sigp || !siglenp)
		return -EINVAL;
	(*sigp) = NULL;
	(*siglenp) = 0;
	len = EVP_PKEY_size(as->pkey);
	sig = malloc(len);
	if (!sig)
		return -ENOMEM;
	EVP_MD_CTX_init(&ctx);
	if (!EVP_SignInit(&ctx, as->type)) {
		EVP_MD_CTX_cleanup(&ctx);
		free(sig);
		return -ENOMEM;
	}
	EVP_SignUpdate(&ctx, buf, buflen);
	result = EVP_SignFinal(&ctx, sig, &len, as->pkey);
	EVP_MD_CTX_cleanup(&ctx);
	if (result != 1) {
		free(sig);
		return -EPERM;
	}
	(*sigp) = sig;
	(*siglenp) = len;
	return 0;
}

int
anoubis_sig_verify_policy(const char *filename, unsigned char *sigbuf,
    int siglen, EVP_PKEY *sigkey)
{
	char		 buffer[1024];
	EVP_MD_CTX	 ctx;
	int		 fd, n, result = 0;

	EVP_MD_CTX_init(&ctx);
	if (EVP_VerifyInit(&ctx, ANOUBIS_SIG_HASH_DEFAULT) == 0) {
		EVP_MD_CTX_cleanup(&ctx);
		return (-5);
	}

	/* Open policy file */
	if ((fd = open(filename, O_RDONLY)) == -1) {
		EVP_MD_CTX_cleanup(&ctx);
		return (-6);
	}

	while ((n = read(fd, buffer, sizeof(buffer))) > 0)
		EVP_VerifyUpdate(&ctx, buffer, n);
	if (n == -1) {
		close(fd);
		EVP_MD_CTX_cleanup(&ctx);
		return (-7);
	}
	close(fd);

	/* Verify */
	if ((result = EVP_VerifyFinal(&ctx, sigbuf, siglen, sigkey)) == -1)
		result = -8;

	EVP_MD_CTX_cleanup(&ctx);

	return (result);
}

unsigned char *
anoubis_sign_policy_buf(struct anoubis_sig *as, char *buf, unsigned int *len)
{
	unsigned char	*sigbuf = NULL;
	unsigned int	 buf_len;
	unsigned int	 siglen = 0;
	EVP_MD_CTX	 ctx;
	EVP_PKEY	*pkey;

	if (as == NULL || len == NULL || buf == NULL) {
		*len = -EINVAL;
		return NULL;
	}

	buf_len = strlen(buf);
	if (buf_len != *len) {
		*len = -ERANGE;
		return NULL;
	}
	pkey = as->pkey;
	siglen = EVP_PKEY_size(pkey);

	if ((sigbuf = calloc(siglen, sizeof(unsigned char))) == NULL)
		return NULL;


	EVP_MD_CTX_init(&ctx);
	if (EVP_SignInit(&ctx, as->type) == 0) {
		EVP_MD_CTX_cleanup(&ctx);
		free(sigbuf);
		*len = -ERR_get_error();
		return NULL;
	}
	EVP_SignUpdate(&ctx, buf, buf_len);
	if (EVP_SignFinal(&ctx, sigbuf, &siglen, pkey) == 0) {
		free(sigbuf);
		EVP_MD_CTX_cleanup(&ctx);
		*len = -ERR_get_error();
		return NULL;
	}
	EVP_MD_CTX_cleanup(&ctx);

	*len = siglen;
	return sigbuf;
}

unsigned char *
anoubis_sign_policy(struct anoubis_sig *as, char *file, unsigned int *len)
{
	unsigned char	*sigbuf = NULL;
	char		*buf[1024];
	int		 fd, n;
	unsigned int	 siglen = 0;
	EVP_MD_CTX	 ctx;
	EVP_PKEY	*pkey;

	if (as == NULL || len == NULL || file == NULL)
		return NULL;

	pkey = as->pkey;
	siglen = EVP_PKEY_size(pkey);

	/* Open Policy file */
	if ((fd = open(file, O_RDONLY)) == -1)
		return NULL;

	if ((sigbuf = calloc(siglen, sizeof(unsigned char))) == NULL) {
		close(fd);
		return NULL;
	}

	EVP_MD_CTX_init(&ctx);
	if (EVP_SignInit(&ctx, as->type) == 0) {
		EVP_MD_CTX_cleanup(&ctx);
		free(sigbuf);
		return NULL;
	}
	while ((n = read(fd, buf, sizeof(buf))) > 0)
		EVP_SignUpdate(&ctx, buf, n);
	if (n == -1) {
		close(fd);
		free(sigbuf);
		EVP_MD_CTX_cleanup(&ctx);
		return NULL;
	}
	close(fd);
	if (EVP_SignFinal(&ctx, sigbuf, &siglen, pkey) == 0) {
		free(sigbuf);
		EVP_MD_CTX_cleanup(&ctx);
		return NULL;
	}
	EVP_MD_CTX_cleanup(&ctx);

	*len = siglen;
	return sigbuf;
}

/* This function will be used inside of the daemon */
int
anoubisd_verify_csum(EVP_PKEY *pkey,
     const unsigned char csum[ANOUBIS_SIG_HASH_SHA256_LEN],
     const unsigned char *sfs_sign, int sfs_len)
{
	int result = 0;
	EVP_MD_CTX ctx;

	if (pkey == NULL || csum == NULL || sfs_sign == NULL)
		return -1;

	EVP_MD_CTX_init(&ctx);
	if (EVP_VerifyInit(&ctx, ANOUBIS_SIG_HASH_DEFAULT) == 0) {
		EVP_MD_CTX_cleanup(&ctx);
		return -1;
	}
	EVP_VerifyUpdate(&ctx, csum, ANOUBIS_SIG_HASH_SHA256_LEN);
	result = EVP_VerifyFinal(&ctx, sfs_sign, sfs_len, pkey);

	EVP_MD_CTX_cleanup(&ctx);

	return result;
}

/**
 * Compare two EVP_PKEYs assuming that they are both RSA keys.
 * The keys are considered to be indentical if public modulus and exponent
 * are the same, i.e. if their public parts are the same.
 * @param key1 The first key.
 * @param key2 The second key.
 * @return Zero if the two keys are equal. -EINVAL if they are different.
 */
static int
anoubis_sig_dorsacmp(EVP_PKEY *key1, EVP_PKEY *key2)
{
	/* XXX CEH: Only works for RSA keys right now. */
	if (key1->type != key2->type || key1->type != EVP_PKEY_RSA)
		return 0;
	if (key1->pkey.rsa->n == NULL || key2->pkey.rsa->n == NULL)
		return -EINVAL;
	if (BN_cmp(key1->pkey.rsa->n, key2->pkey.rsa->n) != 0)
		return -EINVAL;
	if (key1->pkey.rsa->e == NULL || key2->pkey.rsa->e == NULL)
		return -EINVAL;
	if (BN_cmp(key1->pkey.rsa->e, key2->pkey.rsa->e) != 0)
		return -EINVAL;

	return 0;
}

/**
 * Verify that the anoubis signature structure contains consistent
 * public and private keys.
 * @param as The anoubis_sig structure.
 * @return Zero in case of success, a negative error code in case of an error.
 *
 * If either no cert or no private key has been loaded, the comparison will
 * be successful.
 */
int
anoubis_sig_check_consistency(const struct anoubis_sig *as)
{
	EVP_PKEY	*cert_pubkey = NULL;
	int		 ret;

	/* No private key or no certificate. This is ok. */
	if (as->pkey == NULL || as->cert == NULL)
		return 0;
	cert_pubkey = X509_get_pubkey(as->cert);
	if (!cert_pubkey)
		return -EIO;
	ret = anoubis_sig_dorsacmp(cert_pubkey, as->pkey);
	EVP_PKEY_free(cert_pubkey);
	return ret;
}

/**
 * Allocate a new anoubis_sig structure and initialize it with the givenb
 * key and certificate.
 *
 * @param asp The pointer to the new structure is stored *asp.
 * @param keyfile Read the private key from this file.
 * @param certfile Read the certificate from this file.
 * @param passcb Callback function to prompt the user for the password
 *     of the private key.
 * @return A negative error code in case of an error, zero in case of success.
 *     If an error occured *asp will be NULL otherwise *asp will point to
 *     the new anoubis_sig structure.
 *
 * NOTE: This funtion will verify that the private key (if any) matches
 *     the public key extracted from the certificate (if any).
 */
int
anoubis_sig_create(struct anoubis_sig **asp, const char *keyfile,
    const char *certfile, pem_password_cb *passcb)
{

	struct anoubis_sig *as = NULL;
	EVP_PKEY *pkey = NULL;
	FILE *f = NULL;
	X509	*cert = NULL;
	int err = 0;

	OpenSSL_add_all_algorithms();

	if (asp == NULL)
		return -EINVAL;
	*asp = NULL;
	if (!certfile && !keyfile)
		return -EINVAL;
	if (keyfile) {
		f = fopen(keyfile, "r");
		if (f == NULL)
			return -errno;
		pkey = PEM_read_PrivateKey(f, NULL, passcb, &err);
		fclose(f);

		if (!pkey) {
			/* key == NULL and err == 0 implies a bad password */
			if (err == 0)
				return -EPERM;
			/* Handle interesting errors or return -EIO */
			err = ERR_GET_REASON(ERR_peek_last_error());
			ERR_clear_error();

			/* PEM_R_BAD_DECRYPT indicates a bad password */
			if (err == PEM_R_BAD_DECRYPT)
				return -EPERM;
			if (err == PEM_R_BAD_PASSWORD_READ)
				return -ECANCELED;
			return -EIO;
		}
	}
	if (certfile) {
		if ((f = fopen(certfile, "r")) == NULL) {
			err = -errno;
			if (pkey)
				EVP_PKEY_free(pkey);
			return err;
		}
		cert = PEM_read_X509(f, NULL, NULL, NULL);
		fclose(f);
		if (cert == NULL) {
			if (pkey)
				EVP_PKEY_free(pkey);
			return -EINVAL;
		}
	}

	as = calloc(1, sizeof(struct anoubis_sig));
	if (as == NULL) {
		if (pkey)
			EVP_PKEY_free(pkey);
		if (cert)
			X509_free(as->cert);
		return -ENOMEM;
	}
	as->type = ANOUBIS_SIG_HASH_DEFAULT;
	as->pkey = pkey;
	as->cert = cert;
	as->keyid = NULL;
	if (cert) {
		err = cert_keyid(&as->keyid, cert);
		if (err < 0) {
			anoubis_sig_free(as);
			return err;
		}
		as->idlen = err;
	} else {
		as->idlen = 0;
	}
	err = anoubis_sig_check_consistency(as);
	if (err < 0) {
		anoubis_sig_free(as);
		return err;
	}
	(*asp) = as;
	return 0;
}

void
anoubis_sig_free(struct anoubis_sig *as)
{
	if (as == NULL)
		return;

	if (as->pkey)
		EVP_PKEY_free(as->pkey);
	if (as->keyid)
		free(as->keyid);
	if (as->cert)
		X509_free(as->cert);

	EVP_cleanup();
	free(as);
}

char *
anoubis_sig_key2char(int idlen, const unsigned char *keyid)
{
	int i;
	char *res = NULL;

	if ((res = calloc((2*idlen) + 1 , sizeof(char))) == NULL)
		return NULL;

	for (i = 0; i < idlen; i++)
		sprintf(&res[2*i], "%2.2x", keyid[i]);
	res[2*i] = '\0';
	return res;
}

/*
 * NOTE: NEVER write to stdout in this function. Try to use the tty instead.
 */
int
pass_cb(char *buf, int size, int rwflag __used, void *err __used)
{
	int len;
	char *tmp;

	tmp = getpass("Enter passphrase for Private Key: ");
	len = strlen(tmp);
	if (len <= 0)
		return 0;

	if (len > size) len = size;

	memcpy(buf, tmp, len);
	bzero(tmp, len);

	return len;
}

static int
cert_keyid(unsigned char **keyid, X509 *cert)
{
	X509_EXTENSION *ext = NULL;
	ASN1_OCTET_STRING *o_asn = NULL;
	int rc = 0;
	int rv = -1, len = -1;

	if (keyid == NULL || cert == NULL)
		return -EINVAL;

	for (rv = X509_get_ext_by_NID(cert, NID_subject_key_identifier, rv);
	    rv >= 0;
	    rv = X509_get_ext_by_NID(cert, NID_subject_key_identifier, rv))
		ext = X509_get_ext(cert, rv);

	o_asn = X509_EXTENSION_get_data(ext);
	if (!o_asn) {
		rc = ERR_get_error();
		return -rc;
	}
	/* The result of X509_EXTENSION_get_data beginns with
	 * a number which describes algorithm (see RFC 3280)
	 * We just need the KeyID, for that we cut away this number
	 */
	if (!memcmp(o_asn->data, "\004\024", 2)) {
		len = ASN1_STRING_length(o_asn) - 2;
		if ((*keyid = calloc(len, sizeof(unsigned char))) == NULL)
			return -ENOMEM;
		memcpy(*keyid, &o_asn->data[2], len);
	} else {
		len = ASN1_STRING_length(o_asn);
		if ((*keyid = calloc(len, sizeof(unsigned char))) == NULL)
			return -ENOMEM;
		memcpy(*keyid, &o_asn->data, len);
	}

	return (len);
}

char *
anoubis_sig_cert_name(X509 *cert)
{
	BIO	*bio;
	char	*buf = NULL;
	int	num;

	if ((bio = BIO_new(BIO_s_mem())) == NULL)
		return (NULL);

	X509_NAME_print_ex(bio, X509_get_subject_name(cert), 0,
	    XN_FLAG_ONELINE & ~XN_FLAG_SPC_EQ & ~ASN1_STRFLGS_ESC_MSB);

	if ((buf = calloc(ANOUBIS_SIG_NAME_LEN + 1, sizeof(char))) == NULL) {
		BIO_free(bio);
		return (NULL);
	}

	if ((num = BIO_read(bio, buf, ANOUBIS_SIG_NAME_LEN)) < 0) {
		BIO_free(bio);
		free(buf);

		return (NULL);
	}

	buf[num] = '\0';

	BIO_free(bio);

	return (buf);
}

/* The idea of this code is taken from openssl ASN1_UTCTIME_print(). */
#define GETNUM(x) (((x)[0] - '0') * 10 + (x)[1] - '0')
int
anoubis_sig_cert_validity(X509 *cert, struct tm *time)
{
	ASN1_TIME	*notAfter = NULL;

	if ((cert == NULL) || (time == NULL)) {
		return (-EINVAL);
	}

	notAfter = X509_get_notAfter(cert);
	if (notAfter == NULL) {
		return (-EINVAL);
	} else if (notAfter->length < 10 || !ASN1_UTCTIME_check(notAfter)) {
		return (-ERANGE);
	}

	time->tm_year = GETNUM(notAfter->data);
	if (time->tm_year <= 50) {
		time->tm_year += 100;
	}

	/* Type and range checks already done by ASN1_UTCTIME_check(). */
	time->tm_mon  = GETNUM(notAfter->data + 2);
	time->tm_mday = GETNUM(notAfter->data + 4);
	time->tm_hour = GETNUM(notAfter->data + 6);
	time->tm_min  = GETNUM(notAfter->data + 8);
	time->tm_sec  = GETNUM(notAfter->data + 10);

	return (0);
}
#undef GETNUM

/**
 * Check the internal consistency of both keyA and keyB. Additionally,
 * if both signature structures contain are non NULL make sure that they
 * use the same private and public keys.
 *
 * @param keyA The first anoubis_sig structure.
 * @param keyB The second anoubis_sig structure.
 * @return Zero if the comparison succeeds, a negative error code if the
 *     comparison fails.
 */
int
anoubis_sig_keycmp(struct anoubis_sig *keyA, struct anoubis_sig *keyB)
{
	int		 rc;
	EVP_PKEY	*certpubA = NULL, *certpubB = NULL;

	if (keyA && (rc = anoubis_sig_check_consistency(keyA)) < 0)
		return rc;
	if (keyB && (rc = anoubis_sig_check_consistency(keyB)) < 0)
		return rc;
	if (!keyA || !keyB)
		return 0;

	rc = -EIO;
	if (keyA->cert) {
		certpubA = X509_get_pubkey(keyA->cert);
		if (!certpubA)
			goto out;
	}
	if (keyB->cert) {
		certpubB = X509_get_pubkey(keyB->cert);
		if (!certpubB)
			goto out;
	}
	rc = -EPERM;

#define DO_KEY_CMP(X,Y) do {					\
	if ((X) != NULL && (Y) != NULL)				\
		if (anoubis_sig_dorsacmp((X), (Y)) != 0)	\
			goto out;				\
	} while (0)

	DO_KEY_CMP(certpubA, certpubB);
	DO_KEY_CMP(certpubA, keyB->pkey);
	DO_KEY_CMP(keyA->pkey, certpubB);
	DO_KEY_CMP(keyA->pkey, keyB->pkey);
#undef DO_KEY_CMP

	rc = 0;
out:
	if (certpubA)
		EVP_PKEY_free(certpubA);
	if (certpubB)
		EVP_PKEY_free(certpubB);
	return rc;
}
